File: src/option.ts
import { UnwrapError } from './error.ts';
import { Result } from './result.ts';
import { type OptionSerial, readOptionWire } from './wire.ts';

/** A value that is either {@link Some} or {@link None}. */
export type Option<T> = Some<T> | None<T>;

/** Arms for {@link Option} folding. */
export interface OptionMatchArms<T, R> {
  some(value: T): R;
  none(): R;
}

/** Arms for {@link OptionBase.asResult}. */
export interface OptionResultArms<T, R, E> {
  some(value: T): Result<R, E>;
  none(): Result<R, E>;
}

/** Optional encoder for {@link OptionBase.serial}. */
export interface OptionEncode<T> {
  some?(value: T): unknown;
}

/** Decoder for {@link Option.fromSerial}. */
export interface OptionDecode<T> {
  some(value: unknown): T;
}

abstract class OptionBase<T> {
  /** Discriminant tag; `'some'` or `'none'` for a well-formed value. */
  abstract readonly kind: 'some' | 'none';

  protected constructor() {
    /* sealed hierarchy: instances are created only via Some/None */
  }

  /** Whether this option contains a value. */
  get isSome(): boolean {
    return this.kind === 'some';
  }

  /** Whether this option is empty. */
  get isNone(): boolean {
    return this.kind === 'none';
  }

  /**
   * Folds both variants into one return type — the single dispatch point.
   *
   * A value whose tag is neither `'some'` nor `'none'` throws
   * {@link UnwrapError} here rather than returning silently.
   */
  match<R>(arms: OptionMatchArms<T, R>): R {
    if (this.kind === 'some') {
      return arms.some((this as unknown as Some<T>).value);
    }
    if (this.kind === 'none') {
      return arms.none();
    }
    throw new UnwrapError({
      monad: 'option',
      expected: 'Some or None',
      actual: String((this as { kind: unknown }).kind),
    });
  }

  /** Transforms a {@link Some} value and leaves {@link None} untouched. */
  map<R>(fn: (value: T) => R): Option<R> {
    return this.match<Option<R>>({
      some: value => Option.some(fn(value)),
      none: () => Option.none(),
    });
  }

  /** Chains another optional operation after a {@link Some} value. */
  andThen<R>(fn: (value: T) => Option<R>): Option<R> {
    return this.match<Option<R>>({ some: value => fn(value), none: () => Option.none() });
  }

  /** Returns the {@link Some} value or throws {@link UnwrapError}. */
  unwrap(): T {
    return this.match<T>({
      some: value => value,
      none: () => {
        throw new UnwrapError({ monad: 'option', expected: 'Some', actual: 'None' });
      },
    });
  }

  /** Returns the {@link Some} value, or `fallback` for {@link None}. */
  unwrapOr(fallback: T): T {
    return this.match<T>({ some: value => value, none: () => fallback });
  }

  /** Returns the {@link Some} value, or computes a fallback for {@link None}. */
  unwrapOrElse(fn: () => T): T {
    return this.match<T>({ some: value => value, none: () => fn() });
  }

  /** Converts {@link Some} to {@link Ok} and {@link None} to {@link Err}. */
  okOr<E>(error: E): Result<T, E> {
    return this.match<Result<T, E>>({
      some: value => Result.ok(value),
      none: () => Result.err(error),
    });
  }

  /** Maps either option variant directly into a {@link Result}. */
  asResult<R, E>(arms: OptionResultArms<T, R, E>): Result<R, E> {
    return this.match<Result<R, E>>({ some: value => arms.some(value), none: () => arms.none() });
  }

  /** Returns the nullable native representation (`undefined` for {@link None}). */
  native(): T | undefined {
    return this.match<T | undefined>({ some: value => value, none: () => undefined });
  }

  /** Encodes this option using the tagged-object wire contract. */
  serial(encode?: OptionEncode<T>): OptionSerial {
    return this.match<OptionSerial>({
      some: value => ({ kind: 'some', value: encode?.some ? encode.some(value) : value }),
      none: () => ({ kind: 'none' }),
    });
  }
}

/** The populated {@link Option} variant. */
export class Some<T> extends OptionBase<T> {
  readonly kind = 'some' as const;
  readonly value: T;

  constructor(value: T) {
    super();
    this.value = value;
  }
}

/** The empty {@link Option} variant. */
export class None<T> extends OptionBase<T> {
  readonly kind: 'none';

  constructor() {
    super();
    this.kind = 'none';
  }
}

/** Constructors and wire decoder for {@link Option} (const companion). */
export const Option = {
  /** Creates a {@link Some} option. */
  some<T>(value: T): Option<T> {
    return new Some<T>(value);
  },

  /** Creates a {@link None} option. */
  none<T>(): Option<T> {
    return new None<T>();
  },

  /** Converts `null`/`undefined` to {@link None} and every other value to {@link Some}. */
  fromNullable<T>(value: T | null | undefined): Option<T> {
    return value == null ? new None<T>() : new Some<T>(value);
  },

  /** Decodes the `{kind:'some',value} | {kind:'none'}` wire form. */
  fromSerial<T>(serial: OptionSerial, decode: OptionDecode<T>): Option<T> {
    const wire = readOptionWire(serial);
    return wire.kind === 'some' ? Option.some(decode.some(wire.value)) : Option.none();
  },
} as const;

File: Lib/Option.cs
using System.Diagnostics.CodeAnalysis;
using System.Text.Json.Serialization;

namespace AtomiCloud.Diene.Results;

/// <summary>An allocation-free optional value.</summary>
[JsonConverter(typeof(OptionValueJsonConverterFactory))]
public readonly struct Option<T> : IEquatable<Option<T>>
{
    private const byte SomeState = 1;
    private const byte NoneState = 2;
    private readonly byte _state;
    private readonly T? _value;

    internal Option(T value)
    {
        _state = SomeState;
        _value = value;
    }

    internal Option(bool none)
    {
        _state = none ? NoneState : throw new ArgumentOutOfRangeException(nameof(none));
        _value = default;
    }

    /// <summary>Returns whether a value is present.</summary>
    public bool IsSome()
    {
        EnsureInitialized();
        return _state == SomeState;
    }

    /// <summary>Tests for Some and extracts its value.</summary>
    public bool IsSome([MaybeNullWhen(false)] out T value)
    {
        EnsureInitialized();
        value = _value;
        return _state == SomeState;
    }

    /// <summary>Returns whether no value is present.</summary>
    public bool IsNone()
    {
        EnsureInitialized();
        return _state == NoneState;
    }

    /// <summary>Maps the present value.</summary>
    public Option<TOut> Map<TOut>(Func<T, TOut> mapper)
    {
        ArgumentNullException.ThrowIfNull(mapper);
        return IsSome() ? Option.Some(mapper(_value!)) : Option.None<TOut>();
    }

    /// <summary>Chains an Option-returning continuation.</summary>
    public Option<TOut> Then<TOut>(Func<T, Option<TOut>> continuation)
    {
        ArgumentNullException.ThrowIfNull(continuation);
        return IsSome() ? continuation(_value!) : Option.None<TOut>();
    }

    /// <summary>Exhaustively maps either variant.</summary>
    public TOut Match<TOut>(Func<T, TOut> some, Func<TOut> none)
    {
        ArgumentNullException.ThrowIfNull(some);
        ArgumentNullException.ThrowIfNull(none);
        return IsSome() ? some(_value!) : none();
    }

    /// <summary>Gets the present value or throws <see cref="UnwrapException"/>.</summary>
    public T Get()
    {
        EnsureInitialized();
        if (_state == NoneState) throw new UnwrapException("Some", null);
        return _value!;
    }

    /// <summary>Gets the present value or returns a fallback.</summary>
    public T GetOr(T fallback) => IsSome() ? _value! : fallback;

    /// <summary>Gets the present value or computes a fallback.</summary>
    public T GetOr(Func<T> fallback)
    {
        ArgumentNullException.ThrowIfNull(fallback);
        return IsSome() ? _value! : fallback();
    }

    /// <summary>Returns the value or <see langword="null"/>.</summary>
    public T? ToNullable() => IsSome() ? _value : default;

    /// <summary>Converts Some to Success and None to Failure.</summary>
    public Result<T, E> OkOr<E>(E error) => IsSome() ? Result.Ok<T, E>(_value!) : Result.Err<T, E>(error);

    /// <summary>Converts to the explicit C0 wire representation.</summary>
    public OptionSerial<T> ToSerial() => IsSome() ? OptionSerial<T>.Some(_value!) : OptionSerial<T>.None();

    /// <inheritdoc />
    public bool Equals(Option<T> other)
    {
        EnsureInitialized();
        other.EnsureInitialized();
        return _state == other._state && (_state == NoneState || EqualityComparer<T>.Default.Equals(_value, other._value));
    }

    /// <inheritdoc />
    public override bool Equals(object? obj) => obj is Option<T> other && Equals(other);

    /// <inheritdoc />
    public override int GetHashCode()
    {
        EnsureInitialized();
        return HashCode.Combine(_state, _value);
    }

    /// <inheritdoc />
    public override string ToString()
    {
        EnsureInitialized();
        return _state == SomeState ? $"Some({_value})" : "None";
    }

    /// <summary>Determines whether two Options are equal.</summary>
    public static bool operator ==(Option<T> left, Option<T> right) => left.Equals(right);

    /// <summary>Determines whether two Options are unequal.</summary>
    public static bool operator !=(Option<T> left, Option<T> right) => !left.Equals(right);

    private void EnsureInitialized()
    {
        if (_state is not SomeState and not NoneState) throw new InvalidResultException();
    }
}

/// <summary>Factories for Options.</summary>
public static class Option
{
    /// <summary>Creates an Option containing a value.</summary>
    public static Option<T> Some<T>(T value) => new(value);

    /// <summary>Creates an empty Option.</summary>
    public static Option<T> None<T>() => new(true);

    /// <summary>Creates None for null and Some for a non-null value.</summary>
    public static Option<T> FromNullable<T>(T? value) => value is null ? None<T>() : Some(value);

    /// <summary>Creates an Option from its explicit C0 wire representation.</summary>
    public static Option<T> FromSerial<T>(OptionSerial<T> serial)
    {
        ArgumentNullException.ThrowIfNull(serial);
        return serial.Match(Some, None<T>);
    }
}
